### DISCUSSION TITLE
Zconnect 8ch Switch

### DISCUSSION THREAD

**Discussion by dgaust** [zwave-js/zwave-js#4714] [General]
I recently purchased a ZConnect 8CH Switch Module. It is identified by Zwave-JS as a ConnectHome product but there is no device in the database

I grabbed the manual (https://zwave.com.au/wp-content/uploads/2021/10/ZWave_8ch_switch_manual.pdf) and have started to develop a json file for the configuration, but am running into a strange issue when setting the values.

As an example, there are 8 parameters (one for each channel), and there are five values that can be set depending on the  that set the behaviour of the switch depending on the type of switch it is.

For example
```
PARAMETER # 8 - SWITCH # 8 SETTING
Select the type of command that will be sent to the 3th or 11th Association groups, depending on the state of the switch and
its type.
0 - Turn off by pressing / while holding dim down when a monostable switch is connected (Default);
1 - Turn on when pressing / hold down and dim up when a monostable switch is connected.
2 - Bistable button. It has two fixed positions: ON (contact closed) or OFF (contact open).
3 - Scene activation depending on the state of the switch when a monostable switch is connected (see table of scenes). When
this value is selected, it is necessary to add the controller to the 2nd group of links;
4 - Monostable push button switch. Each subsequent press of the button switches the device from 3th group to the opposite
state. Holding down the button will send an up / down dim command to the device from the 11th group.
5 - Bistable push button switch. With its help, you can switch from one possible state to another when changing the position of
the switch button.
```

However, if I set a value at either 4 or 5, it immediately gets set to 3. The below is an extract from the zwavejs2mqtt debug log.

```
{
      "time": "2022-06-15T05:45:11.279Z",
      "event": "value updated",
      "args": [
        {
          "commandClassName": "Configuration",
          "nodeId": 28,
          "commandClass": 112,
          "endpoint": 0,
          "property": 8,
          "newValue": 4,
          "prevValue": 2,
          "propertyName": "Switch 8 Setting"
        }
      ]
    },
    {
      "time": "2022-06-15T05:45:12.333Z",
      "event": "value updated",
      "args": [
        {
          "commandClassName": "Configuration",
          "commandClass": 112,
          "property": 8,
          "newValue": 3,
          "prevValue": 4,
          "propertyName": "Switch 8 Setting"
        }
      ]
    }
```
Would anyone have any suggestions about what I"m doing wrong
[8channelswitch.txt](https://github.com/zwave-js/node-zwave-js/files/8906269/8channelswitch.txt)
?

**Comment by AlCalzone** [comment]
I can't tell 100% without a driver log, but I'd guess that the device does not accept setting the value to 4 or 5. Are you sure you have the correct manual for your firmware version?

**Reply by dgaust** [reply]
Not 100% how to grab the driver log, but I set the logging to debug (in zwavejs2mqtt) and downloaded the log. The offender is node 28.
[zwavejs2mqtt_2022-06-15.log](https://github.com/zwave-js/node-zwave-js/files/8908689/zwavejs2mqtt_2022-06-15.log)

It's difficult to tell re: firmware as it seems to be a particularly niche product and not that well documented. I've emailed the Australian distributor, so if it's the firmware hopefully they have something. As 4 is the value I really need to use.

**Reply by AlCalzone** [reply]
That's not the correct log, but you were close:
https://zwave-js.github.io/zwavejs2mqtt/#/troubleshooting/generating-logs?id=driver-logs

**Reply by dgaust** [reply]
Thanks for the pointer. Hopefully this is the right one (Still node 28)
[zwavejs2mqtt_2022-06-15.log](https://github.com/zwave-js/node-zwave-js/files/8908787/zwavejs2mqtt_2022-06-15.1.log)

I tried two different ways of setting the parameter, using the config option from the json file I made (which is quite possibly wrong)

And the basic configuration that appeared when there was no device json specified. 
<img width="284" alt="image" src="https://user-images.githubusercontent.com/3413862/173816026-b046930f-19ea-4ee2-9c42-4088c0395788.png">

Looking at the logs, it looks like slightly different commands were issued.

```
2022-06-15 21:19:48.282 DEBUG SOCKET: Event ZWAVE_API emitted to QEf1wgnpNnjp8uHdAAAB
2022-06-15 21:19:48.287 INFO ZWAVE: Calling api writeValue with args: [
  { nodeId: 28, commandClass: 112, endpoint: 0, property: 8 },
  4,
  {},
  [length]: 3
]
2022-06-15 21:19:48.289 INFO ZWAVE: Writing 4 to 28-112-0-8
2022-06-15 21:19:54.175 INFO ZWAVE: Node 28: value updated: 112-0-8 3 => 4
2022-06-15 21:19:54.177 INFO ZWAVE: Success zwave api call writeValue true
2022-06-15 21:19:55.222 INFO ZWAVE: Node 28: value updated: 112-0-8 4 => 3
2022-06-15 21:19:56.063 DEBUG ZWAVE: Node 28: mapping value(s) of property 'level' (level) from CC 128 (Battery) to node property 'minBatteryLevel
2022-06-15 21:19:56.064 DEBUG ZWAVE: Node 28: mapping value(s) of property 'level' (level) from CC 128 (Battery) to node property 'batteryLevels
2022-06-15 21:19:56.065 INFO ZWAVE: Node 28: value updated: 128-0-level 80 => 80
2022-06-15 21:19:56.066 INFO ZWAVE: Node 28: value updated: 128-0-isLow false => false
2022-06-15 21:20:01.872 DEBUG SOCKET: Event ZWAVE_API emitted to QEf1wgnpNnjp8uHdAAAB
2022-06-15 21:20:01.877 INFO ZWAVE: Calling api sendCommand with args: [
  { nodeId: 28, commandClass: 112 },
  'set',
  [ 8, 4, 1, [length]: 3 ],
  [length]: 3
]
2022-06-15 21:20:05.897 INFO ZWAVE: Success zwave api call sendCommand undefined
2022-06-15 21:20:08.384 DEBUG ZWAVE: Node 28: mapping value(s) of property 'level' (level) from CC 128 (Battery) to node property 'minBatteryLevel
2022-06-15 21:20:08.384 DEBUG ZWAVE: Node 28: mapping value(s) of property 'level' (level) from CC 128 (Battery) to node property 'batteryLevels
2022-06-15 21:20:08.385 INFO ZWAVE: Node 28: value updated: 128-0-level 80 => 80
2022-06-15 21:20:08.387 INFO ZWAVE: Node 28: value updated: 128-0-isLow false => false
```

**Reply by AlCalzone** [reply]
> Hopefully this is the right one

nope, I can tell by the filename: 
![image](https://user-images.githubusercontent.com/17641229/173829421-68bfd3b2-39d4-4dc6-9738-19d63a85ec56.png)

**Reply by dgaust** [reply]
Yep, that was me skim reading again unfortunately and picking the wrong part of the UI.
[zwavejs_2022-06-16.log](https://github.com/zwave-js/node-zwave-js/files/8914058/zwavejs_2022-06-16.log)

**Comment by dgaust** [comment]
So I have got feedback from the manufacturer that options 4 and 5 do not exist (as AlCalzone noted earlirer) and was mistranslation of the manual.

So using the scene option (3) to try and trigger a scene. To do that, I need to create a group with the right endpoint and the controller, which I have done

![image](https://user-images.githubusercontent.com/3413862/174005961-1b6665ec-5036-4e72-a622-20e554a94f9c.png)

I'm getting a different error now when I trigger the scene, something about about the payload being invalid and the message dropped (could be user error).

```2022-06-16T04:03:55.584Z SERIAL « 0x01090004001c032b015097                                            (11 bytes)
2022-06-16T04:03:55.584Z DRIVER   Dropping message with invalid payload
2022-06-16T04:03:55.584Z DRIVER « [Node 028] [REQ] [ApplicationCommand]
                                  └─[SceneActivationCCSet] [INVALID]
2022-06-16T04:03:55.584Z SERIAL » [ACK]                                                                   (0x06)
2022-06-16T04:03:55.998Z SERIAL « 0x0110000400080a600d0101310504220260f4                              (18 bytes)
2022-06-16T04:03:56.000Z DRIVER   Dropping message because the driver is not ready to handle it yet.
```

[zwavejs_2022-06-16 (2).log](https://github.com/zwave-js/node-zwave-js/files/8915578/zwavejs_2022-06-16.2.log)

**Reply by AlCalzone** [reply]
So, the first command that is being dropped is the Scene Activation command, because it has an invalid format. The field marked with X is missing:
![grafik](https://user-images.githubusercontent.com/17641229/174097911-9733e80c-298e-4190-9882-b2c3b22cb9f7.png)
and it is not optional per the specs:
> This field MUST specify the time that the transition from the current level to the target level (indicated by the Scene ID) should take.

Wonder how this got through certification, since its fairly easy to spot...
This is a bug in the device, but I don't suppose they provide firmware upgrades, do they? I'd appreciate if you could at least ask - if not, I'll add a workaround, I think this one is fairly unproblematic - just ignore the missing duration field.

The second dropped command is a Multilevel Sensor command that just came in too early, so the driver could not decode it yet.

**Reply by dgaust** [reply]
Thanks. There's option exposed in zigbee2mqtt for me to set it

![image](https://user-images.githubusercontent.com/3413862/174163599-41cae05b-261f-42a8-81c6-e92798639e62.png)

I'll give it a shot setting it through there and seeing how it goes. It's not documented well in their manual that there's no default value.

**Reply by dgaust** [reply]
So that didn't work. There is also another configuration option for parameter 9, which could be it (if I think of a poor translation to English)

![image](https://user-images.githubusercontent.com/3413862/174262158-c6ef9501-ba8a-41bb-a6b3-1da7fe2e977c.png)

But setting that doesn't work either and I still get the same error (now Node 29 as I removed the node, removed the config file, and then re-added the device in case my configuration file was the problem).

[zwavejs_2022-06-17 (4).log](https://github.com/zwave-js/node-zwave-js/files/8926152/zwavejs_2022-06-17.4.log)
 
Possibly a firmware issue?

**Reply by AlCalzone** [reply]
Definitely a firmware issue, like I wrote in my last comment 😅

FYI you don't need to re-include nodes to pick up new configuration files. Just restart zwavejs2mqtt and possibly re-interview the node.

**Reply by dgaust** [reply]
Yep, I saw the comment, was just hoping it was user error setting rather than a firmware issue. Cause I can fix user error, not a firmware bug. 

I'll raise it with the supplier, and link to this thread, and hope there's a fix from their end (or as you said before, something you could work around in zwavejs). Thanks.

**Reply by AlCalzone** [reply]
If they don't reply in a sensible timeframe or seem unwilling to do anything about it (despite being non-compliant), let me know. This is one of the minor bugs IMO where a workaround doesn't seem bad.

**Reply by dgaust** [reply]
Unfortunately not even had an acknowledgment from them that they've received my report.

**Reply by AlCalzone** [reply]
https://github.com/zwave-js/node-zwave-js/pull/4748 - this will go into next release probably.

**Reply by dgaust** [reply]
Thank you.

**Reply by AlCalzone** [reply]
It did. 9.4.1, zwavejs2mqtt should follow soon

**Reply by dgaust** [reply]
Thanks, I'll finish off the config file so I can contribute something back, and make a donation to the project. Much appreciated you making the effort.

**Reply by dgaust** [reply]
And just confirming this made it into the zwavejs2mqtt addon for Home Assistant, and the device is now working as intended.
